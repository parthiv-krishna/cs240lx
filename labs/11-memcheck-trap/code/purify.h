#ifndef __PURIFY_H__
#define __PURIFY_H__

#include "src-loc.h"


void purify_init(void);
#define purify_alloc(_n) purify_alloc_raw(_n, SRC_LOC_MK())
#define purify_free(_ptr) purify_free_raw(_ptr, SRC_LOC_MK())

void *purify_alloc_raw(unsigned n, src_loc_t loc);
void purify_free_raw(void *p, src_loc_t loc);

int purify_heap_errors(void);

// just for the debug macro.
#include "memtrace.h"

static inline void purify_verbose_set(int v_p) {
    memtrace_verbose_set(v_p);
}
#endif
