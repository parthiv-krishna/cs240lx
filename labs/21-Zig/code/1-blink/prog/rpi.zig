extern fn nop() void;

// countdown 'ticks' cycles
pub fn delay(ticks: u32) void {
    var cnt = @as(u32, ticks);
    while (cnt > 0) {
        nop();
        cnt -= @as(u32, 1);
    }
}


