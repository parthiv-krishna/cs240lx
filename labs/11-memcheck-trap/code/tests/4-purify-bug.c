
// checks a single memory overflow.
#include "rpi.h"
#include "purify.h"

void notmain(void) {
    trace("should detect use after free\n");
    purify_init();


    enum { N = 6 };
    void *p[N];
    unsigned sz[N];

    for(int i = 0; i < N; i++) {
        sz[i] = rpi_rand32() % 1024;
        p[i] = purify_alloc(sz[i]);
    }

    for(int i = 0; i < N; i++)
        purify_free(p[i]);

    int i = rpi_rand32() % N;
    char *x = p[i];
    unsigned off = sz[i] + rpi_rand32() % 32;
    trace("corrupting [%p] at offset %d [%p]\n", x, off, &x[off]);
    x[off] = 1;
    panic("should not reach here.\n");
}

