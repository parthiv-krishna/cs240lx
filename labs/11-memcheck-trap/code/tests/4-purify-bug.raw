find-ttyusb.c:find_ttyusb:55:FOUND: </dev/ttyUSB0>
opened tty port </dev/ttyUSB0>.
pi-install: tty-usb=</dev/ttyUSB0> program=<./tests/4-purify-bug.bin> about to boot
HASH: simple_boot: sending 28980 bytes, crc32=1298a233
waiting for a start
putting code
pi sent print: <<Flynn Dreilinger Lab 4>: success: Received the program!
>
bootloader: Done.
listening on ttyusb=</dev/ttyUSB0>
TRACE:should detect use after free
TRACE:corrupting [0x182c20] at offset 173 [0x182ccd]
memtrace.c:data_abort_memtrace:100:got a domain section fault at pc=0x8130: address 0x182ccd
TRACE:[pc=0x8130]: store to address 0x182ccd
TRACE:ERROR: illegal store to to FREED block at [pc=0x8130]: [addr=0x182ccd] is 3 bytes after legal mem (block size=171)
TRACE:	logical block id=5, [addr=0x182c20] nbytes=171
TRACE:	Block allocated at: tests/4-purify-bug.c:notmain:17
DONE!!!

Saw done

bootloader: pi exited.  cleaning up
