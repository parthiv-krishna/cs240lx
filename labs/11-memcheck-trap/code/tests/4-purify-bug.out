TRACE:should detect use after free
TRACE:corrupting 
TRACE:: store to address 0x182ccd
TRACE:ERROR: illegal store to to FREED block at  is 3 bytes after legal mem (block size=171)
TRACE:	logical block id=5,  nbytes=171
TRACE:	Block allocated at: tests/4-purify-bug.c:notmain:17
